File: Assignment1/QuestionNo2/QuestionNo2(Basket2).py
# Que: Write a simple Matrix Multiplication program for a given NxN matrix in any two of your preferred Languages from the following listed buckets, where N is iterated through the set of values 64, 128, 256, 512 and 1024. N can either be hardcoded or specified as input. Consider two cases (a) Elements of matrix are of data type Integer and (b) Double In each case, (i.e. Bucket 1 for (a) and (b) + Bucket 2 for (a) and( b)) 

# a. Report the output of the ‘time’ describing the system and CPU times. 
import numpy as np
import time

N_val = [64, 128, 256, 512, 1024]
# N = int(input("Enter value of N (through the set of values 64, 128, 256, 512 and 1024): "))

def matx_multi(A, B, N):
    multi = np.zeros((N, N), dtype=A.dtype)
    for i in range(N):
        for j in range(N):
            for k in range(N):
                multi[i][j] += A[i][k] * B[k][j]
    return multi

for N in N_val:
# a. Matrix Multiplication with Integers
# randomInt = int(input("Enter the maximum value for the matrix elements: "))
    A = np.random.randint(100, size=(N, N))
    B = np.random.randint(100, size=(N, N))

    start_time = time.time()
    multi = matx_multi(A, B, N)
    end_time = time.time()

    int_time_taken = end_time - start_time
    print(f"Matrix multiplication with Integers for N={N} took: {int_time_taken} seconds.")

# b. Matrix Multiplication with Double
# rondomInt = int(input("Enter the maximum value for the matrix elements: "))
    A = np.random.rand(N, N) * 100
    B = np.random.rand(N, N) * 100

    start_time = time.time()
    multi = matx_multi(A, B, N)
    end_time = time.time()

    double_time_taken = end_time - start_time
    print(f"Matrix multiplication with Doubles for N={N} took: {double_time_taken} seconds.")




# b. Using the ‘language hooks’ evaluate the execution time for the meat portions of the program and how much proportion is it w.r.t. total program execution time.
import numpy as np
import time

N_val = [64, 128, 256, 512, 1024]
# N = int(input("Enter value of N (through the set of values 64, 128, 256, 512 and 1024): "))

def matx_multi(A, B, N):
    multi = np.zeros((N, N), dtype=A.dtype)
    for i in range(N):
        for j in range(N):
            for k in range(N):
                multi[i][j] += A[i][k] * B[k][j]
    return multi

print("### Integer Matrix Multiplication ###")
for N in N_val:
    print(f"Matrix size N={N} (Integer elements):")
    # a. Matrix Multiplication with Integers
    # randomInt = int(input("Enter the maximum value for the matrix elements: "))
    total_start = time.perf_counter()

    A = np.random.randint(100, size=(N, N))
    B = np.random.randint(100, size=(N, N))

    meat_start = time.perf_counter()
    multi = matx_multi(A, B, N)
    meat_end = time.perf_counter()

    meat_time = meat_end - meat_start
    total_end = time.perf_counter()
    
    total_time = total_end - total_start

    print(f"  Total time: {total_time} seconds")
    print(f"  Meat portion time (matrix multiplication): {meat_time} seconds")
    print(f"  Proportion of meat portion time: {meat_time / total_time * 100}%\n")



print("### Double Matrix Multiplication ###")
for N in N_val:
    print(f"Matrix size N={N} (Double elements):")
    # b. Matrix Multiplication with Double
    # rondomInt = int(input("Enter the maximum value for the matrix elements: "))

    total_start = time.perf_counter()
    A = np.random.rand(N, N) * 100
    B = np.random.rand(N, N) * 100

    meat_start = time.perf_counter()
    multi = matx_multi(A, B, N)
    meat_end = time.perf_counter()

    meat_time = meat_end - meat_start
    total_end = time.perf_counter()
    
    total_time = total_end - total_start

    print(f"  Total time: {total_time} seconds")
    print(f"  Meat portion time (matrix multiplication): {meat_time} seconds")
    print(f"  Proportion of meat portion time: {meat_time / total_time * 100}%\n")

    start_time = time.time()
    multi = matx_multi(A, B, N)
    end_time = time.time()

    double_time_taken = end_time - start_time
    print(f"Matrix multiplication with Doubles for N={N} took: {double_time_taken} seconds.")

File: Assignment1/QuestionNo2/QuestionNo2(Basket2).c
#I was facing issue in writing c code so I have take a lock of ref here

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void matx_multi(int **A, int **B, int **multi, int N) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            multi[i][j] = 0;
            for (int k = 0; k < N; k++) {
                multi[i][j] += A[i][k] * B[k][j];
            }
        }
    }
}

void matx_multi_double(double **A, double **B, double **multi, int N) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            multi[i][j] = 0.0;
            for (int k = 0; k < N; k++) {
                multi[i][j] += A[i][k] * B[k][j];
            }
        }
    }
}

int** create_matrix(int N, int max_value) {
    int **matrix = (int **)malloc(N * sizeof(int *));
    for (int i = 0; i < N; i++) {
        matrix[i] = (int *)malloc(N * sizeof(int));
        for (int j = 0; j < N; j++) {
            matrix[i][j] = rand() % max_value;
        }
    }
    return matrix;
}

double** create_matrix_double(int N, double max_value) {
    double **matrix = (double **)malloc(N * sizeof(double *));
    for (int i = 0; i < N; i++) {
        matrix[i] = (double *)malloc(N * sizeof(double));
        for (int j = 0; j < N; j++) {
            matrix[i][j] = ((double)rand() / RAND_MAX) * max_value;
        }
    }
    return matrix;
}

void free_matrix(int **matrix, int N) {
    for (int i = 0; i < N; i++) {
        free(matrix[i]);
    }
    free(matrix);
}

void free_matrix_double(double **matrix, int N) {
    for (int i = 0; i < N; i++) {
        free(matrix[i]);
    }
    free(matrix);
}

int main() {
    int N_val[] = {64, 128, 256, 512, 1024};
    int num_sizes = sizeof(N_val) / sizeof(N_val[0]);
    srand(time(NULL));

    printf("### Integer Matrix Multiplication ###\n");
    for (int idx = 0; idx < num_sizes; idx++) {
        int N = N_val[idx];
        printf("Matrix size N=%d (Integer elements):\n", N);

        int **A = create_matrix(N, 100);
        int **B = create_matrix(N, 100);
        int **multi = (int **)malloc(N * sizeof(int *));
        for (int i = 0; i < N; i++) {
            multi[i] = (int *)malloc(N * sizeof(int));
        }

        clock_t total_start = clock();

        clock_t meat_start = clock();
        matx_multi(A, B, multi, N);
        clock_t meat_end = clock();

        clock_t total_end = clock();
        double total_time = (double)(total_end - total_start) / CLOCKS_PER_SEC;
        double meat_time = (double)(meat_end - meat_start) / CLOCKS_PER_SEC;

        printf("  Total time: %f seconds\n", total_time);
        printf("  Meat portion time (matrix multiplication): %f seconds\n", meat_time);
        printf("  Proportion of meat portion time: %f%%\n\n", (meat_time / total_time) * 100);

        free_matrix(A, N);
        free_matrix(B, N);
        free_matrix(multi, N);
    }

    printf("### Double Matrix Multiplication ###\n");
    for (int idx = 0; idx < num_sizes; idx++) {
        int N = N_val[idx];
        printf("Matrix size N=%d (Double elements):\n", N);

        double **A = create_matrix_double(N, 100.0);
        double **B = create_matrix_double(N, 100.0);
        double **multi = (double **)malloc(N * sizeof(double *));
        for (int i = 0; i < N; i++) {
            multi[i] = (double *)malloc(N * sizeof(double));
        }

        clock_t total_start = clock();

        clock_t meat_start = clock();
        matx_multi_double(A, B, multi, N);
        clock_t meat_end = clock();

        clock_t total_end = clock();
        double total_time = (double)(total_end - total_start) / CLOCKS_PER_SEC;
        double meat_time = (double)(meat_end - meat_start) / CLOCKS_PER_SEC;

        printf("  Total time: %f seconds\n", total_time);
        printf("  Meat portion time (matrix multiplication): %f seconds\n", meat_time);
        printf("  Proportion of meat portion time: %f%%\n\n", (meat_time / total_time) * 100);

        free_matrix_double(A, N);
        free_matrix_double(B, N);
        free_matrix_double(multi, N);
    }

    return 0;
}
